Intermittent test failure in AuthenticatedProducerConsumerTest.testAnonymousSyncProducerAndConsumer

```
2019-02-07\T\15:46:31.485 [INFO] Running org.apache.pulsar.client.api.AuthenticatedProducerConsumerTest
2019-02-07\T\15:47:02.036 [ERROR] Tests run: 11, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 30.527 s <<< FAILURE! - in org.apache.pulsar.client.api.AuthenticatedProducerConsumerTest
2019-02-07\T\15:47:02.036 [ERROR] testAnonymousSyncProducerAndConsumer(org.apache.pulsar.client.api.AuthenticatedProducerConsumerTest)  Time elapsed: 1.753 s  <<< FAILURE!
java.util.ConcurrentModificationException
	at java.util.IdentityHashMap$KeySpliterator.tryAdvance(IdentityHashMap.java:1463)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
	at org.apache.pulsar.client.impl.PulsarClientImpl.subscriptionExist(PulsarClientImpl.java:691)
	at org.apache.pulsar.client.impl.PulsarClientImpl.doSingleTopicSubscribeAsync(PulsarClientImpl.java:314)
	at org.apache.pulsar.client.impl.PulsarClientImpl.singleTopicSubscribeAsync(PulsarClientImpl.java:309)
	at org.apache.pulsar.client.impl.PulsarClientImpl.subscribeAsync(PulsarClientImpl.java:286)
	at org.apache.pulsar.client.impl.ConsumerBuilderImpl.subscribeAsync(ConsumerBuilderImpl.java:116)
	at org.apache.pulsar.client.impl.ConsumerBuilderImpl.subscribe(ConsumerBuilderImpl.java:90)
	at org.apache.pulsar.client.api.AuthenticatedProducerConsumerTest.testSyncProducerAndConsumer(AuthenticatedProducerConsumerTest.java:129)
	at org.apache.pulsar.client.api.AuthenticatedProducerConsumerTest.testAnonymousSyncProducerAndConsumer(AuthenticatedProducerConsumerTest.java:249)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java
+++ b/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java
@@ -685,11 +685,13 @@ public class PulsarClientImpl implements PulsarClient {
 
     @SuppressWarnings("unchecked")
     private <T> Optional<ConsumerBase<T>> subscriptionExist(ConsumerConfigurationData<?> conf) {
-        Optional<ConsumerBase<?>> subscriber = consumers.keySet().stream()
-                .filter(consumerBase -> consumerBase.getSubType().equals(PulsarApi.CommandSubscribe.SubType.Shared))
-                .filter(c -> c.getSubscription().equals(conf.getSubscriptionName()))
-                .findFirst();
-        return subscriber.map(ConsumerBase.class::cast);
+        synchronized (consumers) {
+            Optional<ConsumerBase<?>> subscriber = consumers.keySet().stream()
+                    .filter(consumerBase -> consumerBase.getSubType().equals(PulsarApi.CommandSubscribe.SubType.Shared))
+                    .filter(c -> c.getSubscription().equals(conf.getSubscriptionName()))
+                    .findFirst();
+            return subscriber.map(ConsumerBase.class::cast);
+        }
     }
 
     private static EventLoopGroup getEventLoopGroup(ClientConfigurationData conf) {

